llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -4812,7 +4812,14 @@ Instruction *InstCombinerImpl::visitFreeze(FreezeInst &I) {
   //
   // TODO: This could use getBinopAbsorber() / getBinopIdentity() to avoid
   //       duplicating logic for binops at least.
-  auto getUndefReplacement = [&I](Type *Ty) {
+  auto getUndefReplacement = [&I, &AC = this->AC,
+                              &DT = this->DT](Type *Ty) -> Value * {
+    Value *SelectArgument = nullptr;
+    if (match(&I, m_OneUse(m_Freeze(m_Poison()))) &&
+        match(I.user_back(),
+              m_c_Select(m_Specific(&I), m_Value(SelectArgument))) &&
+        isGuaranteedNotToBeUndefOrPoison(SelectArgument, &AC, &I, &DT))
+      return SelectArgument;
```

Comment:
Why is the select being handled separately from the rest? It's not particularly likely that this happens, but conceptually we should handle this exactly the same as the rest, by trying to find the best replacement for all users. This might be multiple selects with the same "other" value. This will also generalize to other patterns where we need to replace with a value rather than constant (say icmp eq/ne, where we'd also want to replace with the other op).

Reply:
The transformation was based on a single use of freeze poison so we hoisted it out of the use analysis loop.

To verify we would want to be able to do something like the following?
https://alive2.llvm.org/ce/z/pyGVuw